### ISSUE TITLE
Regression in variables precedence: extra_vars do not override values inside rolename/vars interpolation

### ISSUE THREAD

**Issue by amyodov** [ansible/ansible#10896]
Upgraded from 1.7.2 to 1.9.0.1, and found the regression in variable interpolation and precedence (still present in 1.9.1): the variables from extra_vars, if used inside an interpolation within another variable defined in <rolename>/vars/main.yml, lose their precedence.

Imagine the following playbook:
## ansible/group_vars/all

```
v1: group-vars-v1
v2: group-vars-v2
```
## ansible/inventories/local

```
[local]
127.0.0.1
```
## ansible/roles/test-role/tasks/main.yml

```
- name: "Print variables"
  debug: msg="v1={{ v1 }}, v2={{ v2 }}, v_total={{ v_total }}"
```
## ansible/roles/test-role/vars/main.yml

```
v_total: "/{{ v1 }}/{{ v2 }}"
```
## ansible/site.yml

```
- name: Test playbook
  hosts: all
  roles:
   - test-role
```

What happens if we launch `ansible-playbook --inventory=ansible/inventories/local --connection=local ansible/site.yml --extra-vars="v1=extra-v1"` (pay attention to `v_total` value):
## Ansible 1.7.2 (correct)

```
PLAY [Test playbook] ********************************************************** 

GATHERING FACTS *************************************************************** 
ok: [127.0.0.1]

TASK: [test-role | Print variables] ******************************************* 
ok: [127.0.0.1] => {
    "msg": "v1=extra-v1, v2=group-vars-v2, v_total=/extra-v1/group-vars-v2"
}

PLAY RECAP ******************************************************************** 
127.0.0.1                  : ok=2    changed=0    unreachable=0    failed=0
```
## Ansible 1.9.0.1/1.9.1 (wrong)

```
PLAY [Test playbook] ********************************************************** 

GATHERING FACTS *************************************************************** 
ok: [127.0.0.1]

TASK: [test-role | Print variables] ******************************************* 
ok: [127.0.0.1] => {
    "msg": "v1=extra-v1, v2=group-vars-v2, v_total=/group-vars-v1/group-vars-v2"
}

PLAY RECAP ******************************************************************** 
127.0.0.1                  : ok=2    changed=0    unreachable=0    failed=0
```

May be related to bug #10817

**Comment by darshanrp** [comment]
Would be great to have this fixed. I'm facing the exact same problem

**Comment by bcoca** [comment]
this probably broke in 1.8 when var merging was restructured.

**Comment by darshanrp** [comment]
Yup, I just upgraded ansible from 1.7.2 to 1.9.1 and started facing this problem. I'm now stuck with 1.9.1 and cannot revert back to 1.7.2
Found a temporary work around by using `set_fact` until we have a fix for this

**Comment by rossimo** [comment]
What is the workaround? I'm trying to fix the same issue.

**Comment by focusaurus** [comment]
@rossimo I had to work around by moving vars from group_vars to `roles/foo/defaults/main.yml`. Not particularly desirable but that's the only thing that seemed to work in my situation.

@darshanrp could you provide more details explaining how you used `set_fact` to work around this?

**Comment by munglerizer** [comment]
ARGH! I have been battling with this issue on my playbook :). Please can it be upgraded from a P3 as it is quite difficult to detect in most playbooks and quite damaging if it does happen.

Also could we expand on the definition in [the docs](https://docs.ansible.com/playbooks_variables.html#variable-precedence-where-should-i-put-a-variable)
`* then comes "most everything else" (command line switches, vars in play, included vars, role vars, etc)`
this is where my problem is occurring with 1.9.1. There must be an order here but as you can't simultaneously aggregate vars in play and role vars (who would win).

**Comment by jimi-c** [comment]
Hi, this situation is no longer present in the devel branch (which will be ansible 2.0). Here is the output using a version of your example above:

<pre>
# ansible-playbook -vv -i inv site.yml --extra-vars="v1=extra-v1" -c local
Using /etc/ansible/ansible.cfg as config file
1 plays in site.yml
PLAY: Test playbook *************************************************************
TASK [setup] ********************************************************************
ok: [127.0.0.1]
TASK [test-role : test-role : Print variables] **********************************
ok: [127.0.0.1] => {
    "changed": false, 
    "msg": "v1=extra-v1, v2=group-vars-v2, v_total=/extra-v1/group-vars-v2"
}
PLAY RECAP **********************************************************************
127.0.0.1                  : ok=2    changed=0    unreachable=0    failed=0   
</pre>

Per our normal policy, we close bug report issues when we see the issue has been corrected in the devel branch, so we will close this issue now.

If you continue seeing any problems related to this issue, or if you have any further questions, please let us know by stopping by one of the two mailing lists, as appropriate:
- https://groups.google.com/forum/#!forum/ansible-project - for user questions, tips, and tricks
- https://groups.google.com/forum/#!forum/ansible-devel - for strategy, future planning, and questions about writing code

Because this project is very active, we're unlikely to see comments made on closed tickets, but the mailing list is a great way to ask questions, or post if you don't think this particular issue is resolved.

Thank you!

**Comment by javiplx** [comment]
That still happens in 1.9.4. Although I understand the 'normal policy', variable precedence is quite different from 2.x to 1.x (among other things), so "update to v2" might not be a solution. This is not going to be solved on stable-1.9? That means that 1.x is considered unmaintained?
I expect variables supplied by the user on command line should override everything (as otherwise is stated in the documentation), so I consider this a major issue.
If this issue is not going to be reopened, please a better hint to '1.8 when var merging was restructured' will make easier to me investigate it.

**Comment by munglerizer** [comment]
Hey, 

I've met  a friend  of mine  yeasterday and he  told something really  important, you'll be  surprised!  Read more about it here <http://modern.studioqoqol.com/e4ksmnju>

Hugs, adam.downer

**Comment by munglerizer** [comment]
Hello! 

Where are  you  going for your  holidays this year? I've  found a nice  holiday camp, just  take a look  <http://order.egyptlawoffice.com/e4hgddgi>

Good wishes, adam.downer

**Comment by etam** [comment]
Looks like it's fixed in 1.9.6
